### PR TITLE
feat(backend): add custom logits processors to unified vLLM and SGLang backends

### PR DESCRIPTION
**Pull request by peilii**
#### Overview:

Follow-up to DIS-2048 (TRT-LLM slice, #10080). Brings the unified custom
logits-processor smoke path (`DYNAMO_ENABLE_TEST_LOGITS_PROCESSOR=1`) to the
**vLLM** and **SGLang** unified backends, implementing §5 and §6 of the
DIS-2048 design and reusing the shared spec layer + `ForcedSequenceLogitsProcessor`
that already landed for TRT-LLM. No Rust / backend-common API or request-shape
changes.

#### Details:

Four commits:

- **`feat(backend)`** — shared serializable wire format
  (`serialize_logits_processor_entries` / `deserialize_logits_processor_entries`)
  so both backends carry per-request activation across their serialization
  boundary (vLLM `extra_args`, SGLang `custom_params`); rejects the
  TRT-LLM-only `PythonProcessorSpec`.
- **`feat(vllm)`** — engine-loaded `DynamoVllmLogitsProcessor(AdapterLogitsProcessor)`,
  registered on `engine_args.logits_processors` at `start()` (only when the hook
  is on and the worker is a generation role), activated per request via
  `SamplingParams.extra_args["dynamo_logits"]`. `new_req_logits_processor`
  realizes fresh per-request state.
- **`feat(sglang)`** — scheduler-side `DynamoSglangLogitProcessor(CustomLogitProcessor)`;
  `from_args` flips `enable_custom_logit_processor` + `skip_tokenizer_init`,
  `generate()` passes the serialized processor + `custom_params`. Per-request
  state is keyed on `context.id()` (unique/stable — not the optional `trace_id`,
  which could collide); a falsy uid is rejected. Forced-sequence only, since
  SGLang's batch callback does not expose per-request token position (arbitrary
  processors deferred per the design).
- **`docs`** — vLLM/SGLang logits-processing pages + nav, and a cross-reference
  from the TRT-LLM page.

The shared generation-stage gating (PREFILL/ENCODE skipped) and per-request
freshness policy are reused from the DIS-2048 layer unchanged.

**Tests / checks run locally:**

- Shared-layer + vLLM-adapter + SGLang-adapter unit tests: **42 passed**
  against `vllm==0.21.0` / `sglang==0.5.12.post1` (CPU; adapter logic only).
- `mypy` clean on changed sources; `pre-commit` (isort/black/flake8/ruff/codespell
  + marker-report) green.
- GPU aggregated smoke (`DYNAMO_ENABLE_TEST_LOGITS_PROCESSOR=1` → response
  contains "Hello world!") to be confirmed in CI / on a GPU box.

**Known limitations / follow-ups (per design):** vLLM/SGLang ship forced-sequence
only; arbitrary `BaseLogitsProcessor` and a public CLI/config loader remain
deferred. The SGLang per-request state relies on the scheduler caching a single
custom-processor instance across decode steps.

#### Where should the reviewer start?

- `components/src/dynamo/common/backend/engine.py` — the serialize/deserialize wire format.
- `components/src/dynamo/sglang/logits_processing/adapter.py` — esp. the per-request
  state assumption (module docstring) and the `context.id()` keying.
- `components/src/dynamo/vllm/logits_processing/adapter.py` + the `vllm`/`sglang`
  `llm_engine.py` `start()` / `generate()` wiring.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- Resolves DIS-2176 (Linear). Follow-up to DIS-2048 / #10080.

---

<a href="https://nvidia.devinenterprise.com/review/ai-dynamo/dynamo/pull/10224" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Logits processing support added for SGLang backend
  * Logits processing support added for vLLM backend
  * Serializable logits processor specifications across request boundaries

* **Tests**
  * Comprehensive unit tests for logits processor runtime behavior in both backends

* **Documentation**
  * Logits processing integration guides for SGLang and vLLM backends

### CODE REVIEW

File: components/src/dynamo/sglang/llm_engine.py [hunk 1]
```diff
@@ -204,6 +230,31 @@ async def start(self, worker_id: int) -> EngineConfig:
...
+        """
+        assert self.engine is not None, "Engine not initialized"
+        tokenizer_manager = getattr(self.engine, "tokenizer_manager", None)
+        tokenizer = getattr(tokenizer_manager, "tokenizer", None)
+        if tokenizer is None:
+            raise RuntimeError(
+                "SGLang engine exposes no tokenizer; "
+                f"{TEST_LOGITS_PROCESSOR_ENV} requires tokenizer init"
+            )
+        return tokenizer
```
[the remark this answers is not in the document]
Comment (peilii) [reply components/src/dynamo/sglang/llm_engine.py:259]: Fixed — `_logits_tokenizer` now raises `RuntimeError("Engine not initialized")` instead of `assert`, matching the tokenizer check just below and the repo guideline on runtime validation.

File: components/src/dynamo/vllm/logits_processing/adapter.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    No-op on empty entries (the common, hook-off case). The engine-loaded
+    :class:`DynamoVllmLogitsProcessor` reads them back per request.
+    """
+    if not entries:
+        return
+    if getattr(sampling_params, "extra_args", None) is None:
+        sampling_params.extra_args = {}
+    sampling_params.extra_args[_DYNAMO_LOGITS_KEY] = serialize_logits_processor_entries(
+        entries
+    )
```
[the remark this answers is not in the document]
Comment (peilii) [reply components/src/dynamo/vllm/logits_processing/adapter.py]: Good catch — `activate_logits_processors` now pops `extra_args["dynamo_logits"]` on the no-activation path (instead of a bare return), so a client-supplied `vllm_xargs["dynamo_logits"]` can't trigger the engine-loaded adapter when shared gating produced no entries. Added a regression test (`test_activate_clears_stale_payload_on_empty_entries`).

File: components/src/dynamo/vllm/tests/test_vllm_logits_runtime.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    activate_logits_processors,
+    register_dynamo_logits_processor,
+)
+
+pytestmark = [
+    pytest.mark.unit,
+    pytest.mark.vllm,
+    pytest.mark.gpu_0,
+    pytest.mark.pre_merge,
+]
```
[the remark this answers is not in the document]
Comment (peilii) [reply components/src/dynamo/vllm/tests/test_vllm_logits_runtime.py:43]: Not applying. The repo's required marker categories (`tests/report_pytest_markers.py` `REQUIRED_CATEGORIES`) are only **Lifecycle / Test Type / Hardware**. This module has `pre_merge` + `unit` + `gpu_0`, satisfying all three — the marker report passes with "Missing sets: 0". `vllm` is a selection marker; there is no required "component"/`core` category (the sibling `test_sglang_decode_handler.py` follows the same set). Resolving.

File: docs/backends/sglang/sglang-logits-processing.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+title: Logits Processing
+---
+
+For general SGLang features and configuration, see the [Reference Guide](sglang-reference-guide.md).
+
+---
+
+Logits processors let you modify the next-token logits at every decoding step (e.g., to apply custom constraints or sampling transforms). Dynamo provides a backend-agnostic interface and an adapter for SGLang so you can plug in custom processors.
+
+### How it works
```
[the remark this answers is not in the document]
Comment (peilii) [reply docs/backends/sglang/sglang-logits-processing.md:13]: Not applying. These pages intentionally mirror the existing `docs/backends/trtllm/trtllm-logits-processing.md`, which uses the same `title:` (h1) → `###` section pattern. Changing only these two would make them inconsistent with the sibling doc, and markdownlint is not enforced in pre-commit. Resolving for consistency.

File: docs/backends/vllm/vllm-logits-processing.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+title: Logits Processing
+---
+
+For general vLLM features and configuration, see the [Reference Guide](vllm-reference-guide.md).
+
+---
+
+Logits processors let you modify the next-token logits at every decoding step (e.g., to apply custom constraints or sampling transforms). Dynamo provides a backend-agnostic interface and an adapter for vLLM so you can plug in custom processors.
+
+### How it works
```
[the remark this answers is not in the document]
Comment (peilii) [reply docs/backends/vllm/vllm-logits-processing.md:13]: Not applying. These pages intentionally mirror the existing `docs/backends/trtllm/trtllm-logits-processing.md`, which uses the same `title:` (h1) → `###` section pattern. Changing only these two would make them inconsistent with the sibling doc, and markdownlint is not enforced in pre-commit. Resolving for consistency.

File: components/src/dynamo/sglang/logits_processing/adapter.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    ) -> Any:
+        if not custom_param_list:
+            return logits
+        for row_idx, params in enumerate(custom_param_list):
+            params = params or {}
+            payload = params.get(_ENTRIES_KEY)
+            if not payload:
+                continue
+            uid = params.get(_UID_KEY)
+            processors = self._per_request.get(uid)
```
Comment (jh-nv) [inline components/src/dynamo/sglang/logits_processing/adapter.py:116]: This key needs to distinguish parallel completions for the same request. `SglangLLMEngine._build_sampling_params()` forwards `n` from the request, but `generate()` passes a single `context.id()` as the UID for all rows of that request. If SGLang expands `n > 1` into multiple batch rows with the same `custom_params`, those rows will all hit the same `ForcedSequenceLogitsProcessor`; row 0 advances the counter, row 1 then starts at the second forced token, etc. That makes parallel completions corrupt each other.

For the env-gated hook, the safest fix is probably to reject/override `n > 1` while logits entries are active, unless SGLang exposes a stable per-completion key that can be included in `custom_params`.

Comment (peilii) [reply components/src/dynamo/sglang/logits_processing/adapter.py:116]: Went with the override — `activate_logits_processors` now forces `n=1` when logits entries are active (warns if `n>1` was requested), so parallel completions can't collide on the shared processor state. Fixed in 36090e46b8.

FWIW SGLang does expose a per-completion handle via `custom_params["__req__"]` → `Req.output_ids` (the way `ThinkingBudgetLogitProcessor` tracks state), so n>1 could be supported properly later — but for this env-gated smoke hook the `n=1` override is the simpler, lower-risk fix. Thanks for catching this!

File: components/src/dynamo/sglang/logits_processing/adapter.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+Per-request state — the central assumption
+-------------------------------------------
+SGLang's callback is ``__call__(logits, custom_param_list)``. It does NOT
+pass the per-request generated-token position, and ``custom_param_list`` is
+static across decode steps. So this adapter only supports
+``ForcedTokenSequenceSpec``, whose realized
+``ForcedSequenceLogitsProcessor`` advances a purely internal counter and
+ignores ``input_ids`` — exactly the narrow, spec-backed case the DIS-2048
+design scoped for SGLang. Arbitrary ``BaseLogitsProcessor`` instances that
```
[the remark this answers is not in the document]
Comment (peilii) [reply components/src/dynamo/sglang/logits_processing/adapter.py]: Removed in 36090e46b8 — dropped the `DIS-2048` reference from the adapter module docstring; the source no longer carries any Linear ticket id (the markdown docs keep theirs under the carve-out).

File: components/src/dynamo/sglang/logits_processing/adapter.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        )
+    requested_n = sampling_params.get("n")
+    if requested_n not in (None, 1):
+        logger.warning(
+            "forcing n=1 for the logits-processor hook (requested n=%s); "
+            "parallel completions share per-request processor state",
+            requested_n,
+        )
+    sampling_params["n"] = 1
+    sampling_params["custom_params"] = {
```
Comment (tanmayv25) [inline components/src/dynamo/sglang/logits_processing/adapter.py]: This is an unconditional assignment to sampling_params["custom_params"]:

```
sampling_params["custom_params"] = {
    _ENTRIES_KEY: serialize_logits_processor_entries(entries),
    _UID_KEY: request_uid,
}
```

If the request (or _build_sampling_params) already populated custom_params for something else, this silently drops it. The vLLM counterpart in components/src/dynamo/vllm/logits_processing/adapter.py:142-144 is careful to preserve existing extra_args (and there's even a regression test for it: test_activate_preserves_existing_extra_args). The SGLang path should do the same — merge into the existing dict instead of replacing it.

Suggested fix:

```
custom_params = sampling_params.setdefault("custom_params", {})
custom_params[_ENTRIES_KEY] = serialize_logits_processor_entries(entries)
custom_params[_UID_KEY] = request_uid
```

File: components/src/dynamo/sglang/logits_processing/adapter.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+history are deliberately out of scope until SGLang exposes that state at this
+callback.
+
+To make the internal counter advance correctly, per-request processor state
+is kept in a dict keyed by a request UID we inject into ``custom_params``.
+This relies on SGLang caching ONE :class:`DynamoSglangLogitProcessor`
+instance (keyed by its serialized string) and reusing it across decode steps
+and requests — true in the SGLang versions we target. The per-UID map is not
+pruned (no completion signal reaches this callback); acceptable for the
+env-gated smoke hook, not a production processor surface.
```
Comment (tanmayv25) [inline components/src/dynamo/sglang/logits_processing/adapter.py:45]: Is this really acceptable? We should mark this as a known issue in our documentation.

File: components/src/dynamo/sglang/llm_engine.py [hunk 10]
```diff
@@ -283,6 +332,23 @@ async def generate(
         sampling_params = self._build_sampling_params(request)
         input_param = self._get_input_param(request)
 
+        # Shared gating returns [] for PREFILL / hook-off (a no-op). Key state
```
Comment (tanmayv25) [inline components/src/dynamo/sglang/llm_engine.py]: I think comments can be simplified. Why talk about trace_id?

File: components/src/dynamo/sglang/llm_engine.py [hunk 11]
```diff
@@ -36,10 +36,15 @@
 from dynamo.common.backend import telemetry
 from dynamo.common.backend.dp_rank import forced_dp_rank, validate_global_dp_rank
 from dynamo.common.backend.engine import (
+    TEST_LOGITS_PROCESSOR_ENV,
```
Comment (tanmayv25) [inline components/src/dynamo/sglang/llm_engine.py]: DYN_ENABLE_TEST_LOGITS_PROCESSOR?

File: components/src/dynamo/sglang/logits_processing/adapter.py (hunk 8 above)
Comment (peilii) [reply components/src/dynamo/sglang/logits_processing/adapter.py]: Fixed in 71c12f6 — `activate_logits_processors` now merges into the existing `custom_params` via `setdefault` instead of replacing it, so a caller-supplied entry isn't dropped. Added a regression test (`test_activate_preserves_existing_custom_params`), mirroring the vLLM `extra_args` one you referenced.

File: components/src/dynamo/sglang/logits_processing/adapter.py (hunk 9 above)
Comment (peilii) [reply components/src/dynamo/sglang/logits_processing/adapter.py:45]: Documented in 71c12f6 — the SGLang logits-processing doc now calls this out explicitly under Current limitations as a **Known issue (unbounded state map)**: the per-request map isn't pruned and grows for the worker's lifetime while the hook is active; fine for the env-gated smoke test, not a production surface.

File: components/src/dynamo/sglang/llm_engine.py (hunk 10 above)
Comment (peilii) [reply components/src/dynamo/sglang/llm_engine.py]: Simplified in 71c12f6 — dropped the trace_id rationale; the comment is now just the gating note (no-op for PREFILL / hook-off).

File: components/src/dynamo/sglang/llm_engine.py (hunk 11 above)
Comment (peilii) [reply components/src/dynamo/sglang/llm_engine.py]: This is the existing `DYNAMO_ENABLE_TEST_LOGITS_PROCESSOR`, shipped in #10080 for the TRT-LLM slice (handler/worker/tests/docs + tests/serve/test_trtllm.py) and shared across all three backends — DIS-2176 deliberately reuses the same hook. Renaming it here would diverge from the merged TRT-LLM env var. Happy to file a separate cross-backend rename to a `DYN_` prefix if you'd prefer that convention project-wide.

Comment (peilii) [reply components/src/dynamo/sglang/llm_engine.py]: Done — renamed to `DYN_ENABLE_TEST_LOGITS_PROCESSOR` in 3800c73952. Both the env-var string and the module constant are now `DYN_ENABLE_TEST_LOGITS_PROCESSOR`, applied across the shared layer and all three backends (TRT-LLM, vLLM, SGLang) plus docs/tests. (Heads-up: this also renames the var on the already-shipped TRT-LLM path, so existing users of `DYNAMO_ENABLE_TEST_LOGITS_PROCESSOR` must switch.)

**Review by dynamo-ops — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]
